python/lsst/meas/astrom/pessimistic_pattern_matcher_b_3D.py

```diff
@@ -0,0 +1,1111 @@
+            cos_rot_sq : float
+                magnitude of the rotation needed to align the two patterns
+                after their centers are shifted on top of each other.
+            shift_matrix : float array
+                3x3 rotation matrix describing the shift needed to align
+                the source and candidate reference center.
+        """
+
+        # Make sure the sine is a real number.
+        if cos_shift > 1.0:
+            cos_shift = 1.
+        elif cos_shift < -1.0:
+            cos_shift = -1.
+        sin_shift = np.sqrt(1 - cos_shift ** 2)
+
+        # If the sine of our shift is zero we only need to use the idenity
+        # matrix for the shift. Else we construct the rotation matrix for
+        # shift.
+        if sin_shift > 0:
+            rot_axis = np.cross(src_center, ref_center)
+            rot_axis /= sin_shift
+
+            rot_cross_matrix = np.array(
+                [[0., -rot_axis[2], rot_axis[1]],
+                 [rot_axis[2], 0., -rot_axis[0]],
+                 [-rot_axis[1], rot_axis[0], 0.]], dtype=np.float64)
+            shift_matrix = (cos_shift*np.identity(3) +
+                            sin_shift*rot_cross_matrix +
+                            (1. - cos_shift)*np.outer(rot_axis, rot_axis))
+        else:
+            shift_matrix = np.identity(3)
+
+        # Now that we have our shift we apply it to the src delta vector
+        # and check the rotation.
+        rot_src_delta = np.dot(shift_matrix, src_delta)
+        cos_rot_sq = (np.dot(rot_src_delta, ref_delta)**2 /
+                      (np.dot(rot_src_delta, rot_src_delta) *
+                       np.dot(ref_delta, ref_delta)))
+        # If the rotation isn't in tolerance return None.
+        if cos_rot_sq < max_cos_rot_sq:
+            return None
+        return pipeBase.Struct(
+            cos_rot_sq=cos_rot_sq,
+            shift_matrix=shift_matrix,)
+
+    def _create_pattern_spokes(self, src_ctr, src_delta_array, src_dist_array,
+                               ref_ctr, ref_ctr_id, ref_delta, ref_dist,
+                               ref_delta_array, ref_dist_array,
+                               ref_id_array, max_dist_rad, n_match):
+        """ Create the indiviual spokes that make up the pattern now that the
+        shift and rotation are within tolerance. If we can't create a valid
+        pattern we exit early.
+
+        Parameters
+        ----------
+        src_ctr : float array
+            3 vector of the source pinwheel center
+        src_delta_array : float array
+            Array of 3 vector deltas between the source center and the pairs
+            that make up the remaining spokes of the pinwheel
+        src_dist_array : float array
+            Array of the distances of each src_delta in the pinwheel
+        ref_ctr : float array
+            3 vector of the candidate refenerce center
+        ref_ctr_id : int
+            id lookup of the ref_ctr into the master reference array
```

Comment:
I think this would be clearer as "ID of the ref_ctr in the master reference array" (if that's what it is).